sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -60,10 +60,41 @@ static const bool UseCopyEngineForD2DCopy = [] {
   return (CopyEngineForD2DCopy && (std::stoi(CopyEngineForD2DCopy) != 0));
 }();
 
+// SYCL_PI_LEVEL_ZERO_USE_COPY_ENGINE can be set to an integer value, or
+// a pair of integer values of the form "lower_index:upper_index".
+// Here, the indices point to copy engines in a list of all available copy
+// engines.
+// This functions returns this pair of indices.
+// If the user specifies only a single integer, a value of 0 indicates that
+// the copy engines will not be used at all. A value of 1 indicates that all
+// available copy engines can be used.
+static const std::pair<int, int> getRangeOfAllowedCopyEngines = [] {
+  const char *EnvVar = std::getenv("SYCL_PI_LEVEL_ZERO_USE_COPY_ENGINE");
+  // If the environment variable is not set, all available copy engines can be
+  // used.
+  if (!EnvVar) {
+    return std::pair<int, int>(0, INT_MAX);
+  }
+  std::string CopyEngineRange = EnvVar;
+  // Environment variable can be a single integer or a pair of integers
+  // separated by ":"
+  auto pos = CopyEngineRange.find(":");
+  if (pos == std::string::npos) {
+    bool UseCopyEngine = (std::stoi(CopyEngineRange) != 0);
+    if (UseCopyEngine)
+      return std::pair<int, int>(0, INT_MAX); // All copy engines can be used.
+    else
+      return std::pair<int, int>(-1, -1); // No copy engines will be used.
+  }
+  int LowerCopyEngineIndex = std::stoi(CopyEngineRange.substr(0, pos));
+  int UpperCopyEngineIndex = std::stoi(CopyEngineRange.substr(pos + 1));
+  return std::pair<int, int>(LowerCopyEngineIndex, UpperCopyEngineIndex);
```

Comment:
Looking at surrounding code it seems like there must be additional check for invalid values.
From my understanding, the requirement is: -1 <= LowerCopyEngineIndex <= UpperCopyEngineIndex <= INT_MAX.

* <= INT_MAX is validated by `std::stoi`, but we must explicitly check other limitations.
* LowerCopyEngineIndex <= UpperCopyEngineIndex is also tested at one place, although I would check it right here.
* -1 <= LowerCopyEngineIndex - is not checked

Reply:
Addressed in next patch. Thanks